design/authenticators/authenticator_management_improvement.md

```diff
@@ -0,0 +1,247 @@
+      - [Functionality](#functionality-1)
+      - [Synopsis](#synopsis-1)
+      - [Example](#example-1)
+    - [Update Authenticator](#update-authenticator)
+      - [Functionality](#functionality-2)
+      - [Synopsis](#synopsis-2)
+      - [Example](#example-2)
+    - [Enable Authenticator](#enable-authenticator)
+      - [Functionality](#functionality-3)
+      - [Synopsis](#synopsis-3)
+      - [Example](#example-3)
+    - [Disable Authenticator](#disable-authenticator)
+      - [Functionality](#functionality-4)
+      - [Synopsis](#synopsis-4)
+      - [Example](#example-4)
+  - [Backwards Compatibility](#backwards-compatibility)
+  - [Performance](#performance)
+  - [Security](#security)
+  - [Documentation](#documentation)
+  - [Drawbacks](#drawbacks)
+  - [Implementation Options](#implementation-options)
+    - [Alternative 1 - Server Side](#alternative-1---server-side)
+      - [Pros](#pros)
+      - [Cons](#cons)
+    - [Alternative 2 - CLI Side](#alternative-2---cli-side)
+      - [Pros](#pros-1)
+      - [Cons](#cons-1)
+  - [Version Update](#version-update)
+  - [Delivery Plan](#delivery-plan)
+  
+## Introduction
+
+Conjur authenticators are defined as sub-policies, under the `conjur` policy.  
+Currently to define an authenticator, the user needs to complete the following steps:
+
+- Read the documentation
+- Understand how the authenticator policy is structured
+- Create a policy yaml file that defines the authenticator
+- Load the policy using Conjur CLI
+- Enable the authenticator through an environment variable or REST API call
+
+To delete an authenticator, the user needs to complete these steps.  
+Either:
+
+- Create a policy with a `- !delete` statement, on the authenticator policy
+- Load the policy using Conjur CLI
+
+Or:
+
+- Update the `conjur` policy yaml file and take out the authenticator
+- Reload the `conjur` policy using the `--replace` option  
+
+The current user experience requires a certain level of expertise. The user needs to understand the policy structure of each authenticator and the process is comprised of multiple steps. This proposal offers an easier way for managing authenticators.
+
+## Proposed Solution
+
+The proposed solution is to add a few simple Conjur CLI commands (and possibly APIs, see two alternatives below), that would hide the complexity and provide a single step to perform each action:
+
+### Add Authenticator
+
+#### Functionality
+
+Defines a new authenticator in Conjur. Behind the scenes, this command will creat the authenticator policy and fill it with all its required data. If the data is not required to be given by the user, like the `authn-k8s` CA cert and key, the information will be generated automatically.
+
+#### Synopsis
+
+conjur config authenticator add *authenticator-type/service-id* [*parameters*]
+
+#### Example
+
+```shell script
+$ conjur config authenticator add authn-oidc/my-idp --provider-uri my-uri --id-token-user-property prop
```

Comment:
I think that `--provider-uri` is quite error prone as you assume that the only addition that will be is for variables. I'd rather have something like `--metadata provider-uri=my-uri`.

This way you have the flexibility of adding other authenticator parameters in the future and you can encapsulate the step for populating the variable.

Reply:
I don't understand why it's error prone and how your suggestion solves it. Can you please elaborate?
The reason why I suggested authenticator specific flags, is to guide the user to provide the mandatory information, instead of requiring the users to check the docs and understand by themselves what information needs to be provided

Reply:
what i meant is that if authenticators will have other members than variables then this won't work. However, if you use specific tags then it's ok because the implementation is on our side.

I wouldn't go in that direction though as it's can't be generic and will require implementation for each authenticator (with the need to fix it when we do changes)

Reply:
you suggest that if the `--provider-uri` flag is not present then we will fail the request?

Reply:
Yes, the user experience would be that if you don't specify the required parameters it will fail.
Please have a look at my latest addition, based on the suggestion that Srdjan raised. I think this solves your concern as I understand it.